BUGFIX: Graceful error handling if setup not completed

mhsdesign opened the pull request:
Resolves: https://github.com/neos/neos-development-collection/issues/4714
Resolves: https://github.com/neos/neos-development-collection/issues/4446

Current when visiting `http://127.0.0.1:8081` or `http://127.0.0.1:8081/anythingunmatched` without a neos site being imported:

(yes a white-screen if a site exists but no site node)
<img width="825" alt="image" src="https://github.com/neos/neos-development-collection/assets/85400359/f6282fea-88eb-423f-9766-fa3760b03dff">

(or if you have not site the `EventSourcedFrontendNodeRoutePartHandler`)
<img width="1070" alt="image" src="https://github.com/neos/neos-development-collection/assets/85400359/974bea40-065a-4e28-8556-64a276fdc3ca">


Proposed:

<img width="730" alt="image" src="https://github.com/neos/neos-development-collection/assets/85400359/6ef4c5cc-c085-40c1-9053-4f42e61ff88c">


As soon as a site exist, the fusion exception view will use the `error` path as usual to render your custom 404 page.

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign commented:
@kitsunet that should fix most cases and fix the flow tests, ~but im not sure this is all perfect. Seems a bit hacked together, and works only well in exactly this alignment~

With https://github.com/neos/neos-development-collection/pull/4880 its nearly perfect :)

mhsdesign committed:
TASK: Improve error message if node address from router is stale

The router cache is still used in combination with our routing projection.
That could possibly lead to this error when our cache flushing cant keep up, like when deleting tables :D

> TODO: SITE NOT FOUND; should not happen (for address NodeAddress[...]

This changes the error to point to a possible cause, the Flow_Mvc_Routing_Route

Ideally our `RouterCacheHook` should always keep up.

mhsdesign committed:
TASK: Improve error message if site entity exists but no cr

mhsdesign committed:
TASK: Improve error message if the site entity is to be removed but the cr is not setup

mhsdesign committed:
BUGFIX: Prevent the Fusion Exception view to return blank page

mhsdesign committed:
TASK: Remove unused `NoHomepageException`

mhsdesign committed:
TASK: Throw dedicated `SiteDetectionFailedException` exception

mhsdesign committed:
BUGFIX: Graceful site detection and early return

mhsdesign committed:
BUGFIX: Use the welcome view if the SiteDetectionFailed

mhsdesign commented:
This pr makes https://github.com/neos/neos-development-collection/pull/4050 obsolete.

mhsdesign committed:
BUGFIX: Allow standalone instance of `FusionView` without request

previously the `@package` resolving would always take place, even if not necessary, when all includes are fully specified.

mhsdesign committed:
TASK: Followup #4880 use fusion view instead of fluid for welcome rendering fallback

mhsdesign committed:
TASK: Remove deprecated `editPreviewMode` variable

The TypoScript context variable `editPreviewMode` was deprecated in favour of `${documentNode.context.currentRenderingMode.name}`

 With neos 9 that again was removed in favour of the fusion global `renderingMode`.

https://github.com/neos/neos-development-collection/pull/4505

mhsdesign committed:
TASK: Fix phpstan

mhsdesign commented:
The Flow functional tests pass now when Neos is installed. So under that measurement the change is correct :D

bwaidelich commented on Neos.Neos/Classes/View/FusionExceptionView.php:
```diff
@@ -91,7 +92,7 @@ class FusionExceptionView extends AbstractView
 
     /**
-     * @return string
+     * @return mixed
```
nitpick: is this really `mixed` or can we have a more specific return type`

mhsdesign replied:
yeah ... i fiddled around, truth is the `FusionView` for welcomeTemplate returns `mixed` thats why

bwaidelich reviewed (approved):
Nice, +1 by reading

mhsdesign commented on Neos.Neos/Configuration/Settings.yaml:
```diff
@@ -481,7 +481,7 @@ Neos:
 
           noHomepageException:
             matchingExceptionClassNames:
-              - Neos\Neos\Routing\Exception\NoHomepageException
+              - Neos\Neos\FrontendRouting\SiteDetection\SiteDetectionFailedException
             options: *welcomeTemplate
```
i have to evaluate if a `SiteDetectionFailedException` will with my adjustments even come that far ;)

mhsdesign replied:
Yes the only case where this currently happens, is at the `neos/content` route, as the backend controller cannot find a site.

But in 8.3 we dont have a graceful error as well only something like `Site doest not exist for current user`: https://github.com/neos/neos-ui/issues/3182

We should probably redirect to some other available module, or at least keep the backend frame around and not destroy everything.

mhsdesign replied:
i will remove the `noHomepageException` rendering group as this should be solve differently see https://github.com/neos/neos-ui/issues/3721

mhsdesign committed:
TASK: Remove `noHomepageException` rendering group

After the previous adjustments the `SiteDetectionFailedException` will only be thrown in fewer cases, the only real reproducible the `/neos/content` route.
The rendering group is a bit overkill only for this thing and also might swallow 500 errors that should be presented as is and not wrapped by a vibrant welcome page.

The ui should be adjusted instead. https://github.com/neos/neos-ui/issues/3721

mhsdesign committed:
TASK: Docs `SiteDetectionFailedException`

mhsdesign committed:
TASK: Use tabs in xlf??
